narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,233 +3,404 @@
+                .read_last_committed_round(authority)?
+                .unwrap_or(0)),
+            None => Ok(self
+                .certificate_store
+                .last_round_number(authority)?
+                .unwrap_or(0)),
         }
     }
 
-    fn update_metrics(&self, waiting_len: usize) {
-        self.metrics
-            .pending_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(self.pending.len() as i64);
-        self.metrics
-            .waiting_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(waiting_len as i64);
+    #[allow(clippy::mutable_key_type)]
+    fn all_committed_rounds(&self) -> DagResult<BTreeMap<PublicKey, Round>> {
+        let mut all_committed_rounds = BTreeMap::new();
+        for (name, _) in self.committee.authorities() {
+            // Last round is 0 (genesis) when authority is not found in store.
+            let last_round = self.last_committed_round(name)?;
+            all_committed_rounds.insert(name.clone(), last_round);
+        }
+        Ok(all_committed_rounds)
     }
 }
+
+#[allow(clippy::mutable_key_type)]
+async fn run_fetch_task(
+    state: Arc<CertificateWaiterState>,
+    committee: Committee,
+    committed_rounds: BTreeMap<PublicKey, Round>,
+) -> DagResult<()> {
+    // Send request to fetch certificates.
+    let request = FetchCertificatesRequest {
+        exclusive_lower_bounds: committed_rounds.into_iter().collect(),
+        max_items: MAX_CERTIFICATES_TO_FETCH,
+    };
+    let response =
+        fetch_certificates_helper(&state.name, &state.network, &committee, request).await;
+
+    // Process and store fetched certificates.
+    let num_certs_fetched = response.certificates.len();
+    process_certificates_helper(response, &state.tx_certificates_loopback).await?;
+    state
+        .metrics
+        .certificate_waiter_num_certificates_processed
+        .with_label_values(&[&committee.epoch().to_string()])
+        .add(num_certs_fetched as i64);
+
+    debug!("Successfully fetched and processed {num_certs_fetched} certificates");
+    Ok(())
+}
+
+/// Fetches certificates from other primaries concurrently, with ~5 sec interval between each request.
+/// Terminates after the 1st successful response is received.
+#[instrument(level = "debug", skip_all)]
+async fn fetch_certificates_helper(
+    name: &PublicKey,
+    network: &P2pNetwork,
+    committee: &Committee,
+    request: FetchCertificatesRequest,
+) -> FetchCertificatesResponse {
+    trace!("Start sending fetch certificates requests");
+    let request_interval = Duration::from_secs(5);
+    let mut peers: Vec<NetworkPublicKey> = committee
+        .others_primaries(name)
+        .into_iter()
+        .map(|(_, _, network_key)| network_key)
+        .collect();
+    loop {
+        peers.shuffle(&mut ThreadRng::default());
```

Comment:
Makes total sense. I'm think of ranking them by stake weight, divide into 3~5 groups, and then randomize for load balancing in case many authorities start to fetch around the same time? Added a TODO.

Reply:
Nodes with more stake should have more resources, and receive more rewards for them. Trust me, if you do not sample by stake the security arguments breaks down, since a very large number of validators with tiny tiny stake can waste a lot of your time.

Reply:
Thanks @gdanezis for the insights on this, got it! Ok then I see this as an axiom we should use for other cases as well (example sync missing batches)